# **Product Recommendation System** #Thanh Duong

In [2]:
import pandas as pd

**# 1. Load Data**

In [3]:
#df = pd.read_excel(io='../data/Online_Retail.xlsx', sheet_name='Online_Retail')
df = pd.read_excel(io='/content/drive/My Drive/Colab Notebooks/data/Online_Retail.xlsx', sheet_name='Online_Retail')
df.shape

(541909, 8)

In [ ]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [ ]:
df. InvoiceDate

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
541904   2011-12-09 12:50:00
541905   2011-12-09 12:50:00
541906   2011-12-09 12:50:00
541907   2011-12-09 12:50:00
541908   2011-12-09 12:50:00
Name: InvoiceDate, Length: 541909, dtype: datetime64[ns]

# **2. Data Preparation**

## -Handle NaNs in CustomerID field



In [ ]:
df['CustomerID'].describe()

count    406829.000000
mean      15287.690570
std        1713.600303
min       12346.000000
25%       13953.000000
50%       15152.000000
75%       16791.000000
max       18287.000000
Name: CustomerID, dtype: float64

In [ ]:
df['CustomerID'].isna().sum() # Đếm số bản ghi không có mã KH

135080

In [ ]:
df.loc[df['CustomerID'].isna()].head()  # Lọc bản ghi không có mã KH

InvoiceNo StockCode                      Description  Quantity  \
622     536414     22139                              NaN        56   
1443    536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444    536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445    536544     21786               POLKADOT RAIN HAT          4   
1446    536544     21787            RAIN PONCHO RETROSPOT         2   

             InvoiceDate  UnitPrice  CustomerID         Country  
622  2010-12-01 11:52:00       0.00         NaN  United Kingdom  
1443 2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1444 2010-12-01 14:32:00       2.51         NaN  United Kingdom  
1445 2010-12-01 14:32:00       0.85         NaN  United Kingdom  
1446 2010-12-01 14:32:00       1.66         NaN  United Kingdom

In [ ]:
df.shape

(541909, 8)

In [ ]:
df = df.dropna(subset=['CustomerID']) # Xóa bản ghi có mã KH trống

In [ ]:
df.shape

(406829, 8)

In [ ]:
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

# **- Customer-Item Matrix**

In [ ]:
customer_item_matrix = df.pivot_table(
    index='CustomerID',
    columns='StockCode',
    values='Quantity',
    aggfunc='sum'
)

In [ ]:
customer_item_matrix.loc[12481:].head()

StockCode   10002  10080  10120  10125  10133  10135  11001  15030  15034  \
CustomerID                                                                  
12481.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12483.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   
12484.0       NaN    NaN    NaN    NaN    NaN    NaN   16.0    NaN    NaN   
12488.0       NaN    NaN    NaN    NaN    NaN   10.0    NaN    NaN    NaN   
12489.0       NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN    NaN   

StockCode   15036  ...  90214Y  90214Z  BANK CHARGES  C2  CRUK   D  DOT   M  \
CustomerID         ...                                                        
12481.0      36.0  ...     NaN     NaN           NaN NaN   NaN NaN  NaN NaN   
12483.0       NaN  ...     NaN     NaN           NaN NaN   NaN NaN  NaN NaN   
12484.0       NaN  ...     NaN     NaN           NaN NaN   NaN NaN  NaN NaN   
12488.0       NaN  ...     NaN     NaN           NaN NaN   NaN NaN  NaN NaN   
12489.0       NaN  ...     NaN     NaN           NaN NaN   NaN NaN  NaN NaN   

StockCode   PADS  POST  
CustomerID              
12481.0      NaN  32.0  
12483.0      NaN  13.0  
12484.0      NaN  21.0  
12488.0      NaN   3.0  
12489.0      NaN   2.0  

[5 rows x 3684 columns]

In [ ]:
customer_item_matrix.shape

(4372, 3684)

In [ ]:
df['StockCode'].nunique()

3684

In [ ]:
df['CustomerID'].nunique()

4372

In [ ]:
customer_item_matrix.loc[12348.0].sum()

2341.0

In [ ]:
customer_item_matrix = customer_item_matrix.applymap(lambda x: 1 if x > 0 else 0)

In [ ]:
customer_item_matrix.loc[12481:].head()

StockCode   10002  10080  10120  10125  10133  10135  11001  15030  15034  \
CustomerID                                                                  
12481.0         0      0      0      0      0      0      0      0      0   
12483.0         0      0      0      0      0      0      0      0      0   
12484.0         0      0      0      0      0      0      1      0      0   
12488.0         0      0      0      0      0      1      0      0      0   
12489.0         0      0      0      0      0      0      0      0      0   

StockCode   15036  ...  90214Y  90214Z  BANK CHARGES  C2  CRUK  D  DOT  M  \
CustomerID         ...                                                      
12481.0         1  ...       0       0             0   0     0  0    0  0   
12483.0         0  ...       0       0             0   0     0  0    0  0   
12484.0         0  ...       0       0             0   0     0  0    0  0   
12488.0         0  ...       0       0             0   0     0  0    0  0   
12489.0         0  ...       0       0             0   0     0  0    0  0   

StockCode   PADS  POST  
CustomerID              
12481.0        0     1  
12483.0        0     1  
12484.0        0     1  
12488.0        0     1  
12489.0        0     1  

[5 rows x 3684 columns]

# **3. Collaborative Filtering**

In [ ]:
from sklearn.metrics.pairwise import cosine_similarity

## **3.1. User-based Collaborative Filtering**

## - User-to-User Similarity Matrix



In [ ]:
user_user_sim_matrix = pd.DataFrame(
    cosine_similarity(customer_item_matrix)
)

In [ ]:
user_user_sim_matrix.head()

0         1         2         3         4         5     6         7     \
0   0.0  0.000000  0.000000  0.000000  0.000000  0.000000   0.0  0.000000   
1   0.0  1.000000  0.063022  0.046130  0.047795  0.038814   0.0  0.025876   
2   0.0  0.063022  1.000000  0.024953  0.051709  0.027995   0.0  0.027995   
3   0.0  0.046130  0.024953  1.000000  0.056773  0.138314   0.0  0.030737   
4   0.0  0.047795  0.051709  0.056773  1.000000  0.031846   0.0  0.000000   

       8         9     ...  4362  4363      4364  4365      4366      4367  \
0  0.000000  0.000000  ...   0.0   0.0  0.000000   0.0  0.000000  0.000000   
1  0.136641  0.094742  ...   0.0   0.0  0.054656   0.0  0.032844  0.062318   
2  0.118262  0.146427  ...   0.0   0.0  0.118262   0.0  0.000000  0.000000   
3  0.032461  0.144692  ...   0.0   0.0  0.000000   0.0  0.039014  0.000000   
4  0.000000  0.033315  ...   0.0   0.0  0.000000   0.0  0.000000  0.000000   

   4368      4369      4370      4371  
0   0.0  0.000000  0.000000  0.000000  
1   0.0  0.113776  0.109364  0.012828  
2   0.0  0.000000  0.170905  0.083269  
3   0.0  0.067574  0.137124  0.030475  
4   0.0  0.000000  0.044866  0.000000  

[5 rows x 4372 columns]

In [ ]:
user_user_sim_matrix.columns = customer_item_matrix.index

user_user_sim_matrix['CustomerID'] = customer_item_matrix.index
user_user_sim_matrix = user_user_sim_matrix.set_index('CustomerID')

In [ ]:
user_user_sim_matrix.head()

CustomerID  12346.0   12347.0   12348.0   12349.0   12350.0   12352.0  \
CustomerID                                                              
12346.0         0.0  0.000000  0.000000  0.000000  0.000000  0.000000   
12347.0         0.0  1.000000  0.063022  0.046130  0.047795  0.038814   
12348.0         0.0  0.063022  1.000000  0.024953  0.051709  0.027995   
12349.0         0.0  0.046130  0.024953  1.000000  0.056773  0.138314   
12350.0         0.0  0.047795  0.051709  0.056773  1.000000  0.031846   

CustomerID  12353.0   12354.0   12355.0   12356.0  ...  18273.0  18274.0  \
CustomerID                                         ...                     
12346.0         0.0  0.000000  0.000000  0.000000  ...      0.0      0.0   
12347.0         0.0  0.025876  0.136641  0.094742  ...      0.0      0.0   
12348.0         0.0  0.027995  0.118262  0.146427  ...      0.0      0.0   
12349.0         0.0  0.030737  0.032461  0.144692  ...      0.0      0.0   
12350.0         0.0  0.000000  0.000000  0.033315  ...      0.0      0.0   

CustomerID   18276.0  18277.0   18278.0   18280.0  18281.0   18282.0  \
CustomerID                                                             
12346.0     0.000000      0.0  0.000000  0.000000      0.0  0.000000   
12347.0     0.054656      0.0  0.032844  0.062318      0.0  0.113776   
12348.0     0.118262      0.0  0.000000  0.000000      0.0  0.000000   
12349.0     0.000000      0.0  0.039014  0.000000      0.0  0.067574   
12350.0     0.000000      0.0  0.000000  0.000000      0.0  0.000000   

CustomerID   18283.0   18287.0  
CustomerID                      
12346.0     0.000000  0.000000  
12347.0     0.109364  0.012828  
12348.0     0.170905  0.083269  
12349.0     0.137124  0.030475  
12350.0     0.044866  0.000000  

[5 rows x 4372 columns]

# **- Making Recommendations**

In [ ]:
user_user_sim_matrix.loc[12350.0].sort_values(ascending=False) #Xếp hạng những KH có độ đo tương tự với KH 12350

CustomerID
12350.0    1.000000
17935.0    0.183340
12414.0    0.181902
12652.0    0.175035
12814.0    0.171499
             ...   
14869.0    0.000000
14870.0    0.000000
14873.0    0.000000
14875.0    0.000000
18287.0    0.000000
Name: 12350.0, Length: 4372, dtype: float64

In [ ]:
# Lọc những mặt hàng mà khách hàng A mã 12350.0 đã mua
items_bought_by_A = set(customer_item_matrix.loc[12350.0].iloc[
    customer_item_matrix.loc[12350.0].to_numpy().nonzero()].index)
items_bought_by_A

{20615,
 20652,
 21171,
 21832,
 21864,
 21866,
 21908,
 21915,
 22348,
 22412,
 22551,
 22557,
 22620,
 '79066K',
 '79191C',
 '84086C',
 'POST'}

In [ ]:
# Lọc những mặt hàng mà khách hàng B mã 17935.0 đã mua
items_bought_by_B = set(customer_item_matrix.loc[17935.0].iloc[
    customer_item_matrix.loc[17935.0].to_numpy().nonzero()
].index)
items_bought_by_B

{20657,
 20659,
 20828,
 20856,
 21051,
 21866,
 21867,
 22208,
 22209,
 22210,
 22211,
 22449,
 22450,
 22551,
 22553,
 22557,
 22640,
 22659,
 22749,
 22752,
 22753,
 22754,
 22755,
 23290,
 23292,
 23309,
 '85099B',
 'POST'}

In [ ]:
#Danh sách mã các mặt hàng gợi ý cho khách hàng B mã 17935.0
items_to_recommend_to_B = items_bought_by_A - items_bought_by_B
items_to_recommend_to_B

{20615,
 20652,
 21171,
 21832,
 21864,
 21908,
 21915,
 22348,
 22412,
 22620,
 '79066K',
 '79191C',
 '84086C'}

In [ ]:
#Danh sách mã và tên các mặt hàng gợi ý cho khách hàng B mã 17935.0
df.loc[
    df['StockCode'].isin(items_to_recommend_to_B),
    ['StockCode', 'Description']
].drop_duplicates().set_index('StockCode')

Description
StockCode                                 
21832                 CHOCOLATE CALCULATOR
21915               RED  HARMONICA IN BOX 
22620          4 TRADITIONAL SPINNING TOPS
79066K                      RETRO MOD TRAY
21864      UNION JACK FLAG PASSPORT COVER 
79191C         RETRO PLASTIC ELEPHANT TRAY
21908        CHOCOLATE THIS WAY METAL SIGN
20615         BLUE POLKADOT PASSPORT COVER
20652           BLUE POLKADOT LUGGAGE TAG 
22348          TEA BAG PLATE RED RETROSPOT
22412      METAL SIGN NEIGHBOURHOOD WITCH 
21171                 BATHROOM METAL SIGN 
84086C             PINK/PURPLE RETRO RADIO

# **3.2. Item-based Collaborative Filtering**

---
- Item-to-Item Similarity Matrix


In [ ]:
item_item_sim_matrix = pd.DataFrame(cosine_similarity(customer_item_matrix.T))

In [ ]:
item_item_sim_matrix.columns = customer_item_matrix.T.index

item_item_sim_matrix['StockCode'] = customer_item_matrix.T.index
item_item_sim_matrix = item_item_sim_matrix.set_index('StockCode')

In [ ]:
item_item_sim_matrix

StockCode     10002     10080     10120     10125     10133     10135  \
StockCode                                                               
10002      1.000000  0.000000  0.094868  0.090351  0.063246  0.098907   
10080      0.000000  1.000000  0.000000  0.032774  0.045883  0.047836   
10120      0.094868  0.000000  1.000000  0.057143  0.060000  0.041703   
10125      0.090351  0.032774  0.057143  1.000000  0.042857  0.044682   
10133      0.063246  0.045883  0.060000  0.042857  1.000000  0.281494   
...             ...       ...       ...       ...       ...       ...   
D          0.000000  0.000000  0.000000  0.000000  0.000000  0.000000   
DOT        0.000000  0.000000  0.000000  0.000000  0.000000  0.104257   
M          0.059423  0.017244  0.075165  0.075165  0.067648  0.054855   
PADS       0.000000  0.000000  0.000000  0.000000  0.050000  0.000000   
POST       0.070057  0.000000  0.000000  0.071209  0.016615  0.028871   

StockCode     11001     15030     15034     15036  ...    90214Y  90214Z  \
StockCode                                          ...                     
10002      0.095346  0.047673  0.076139  0.091050  ...  0.000000     0.0   
10080      0.000000  0.000000  0.082855  0.049541  ...  0.000000     0.0   
10120      0.060302  0.060302  0.096309  0.028793  ...  0.000000     0.0   
10125      0.043073  0.000000  0.051594  0.030849  ...  0.000000     0.0   
10133      0.045227  0.060302  0.072232  0.057585  ...  0.000000     0.0   
...             ...       ...       ...       ...  ...       ...     ...   
D          0.000000  0.000000  0.000000  0.000000  ...  0.000000     0.0   
DOT        0.150756  0.000000  0.000000  0.000000  ...  0.000000     0.0   
M          0.101983  0.067989  0.063341  0.075747  ...  0.000000     0.0   
PADS       0.000000  0.000000  0.000000  0.000000  ...  0.000000     0.0   
POST       0.050098  0.000000  0.020003  0.043854  ...  0.031976     0.0   

StockCode  BANK CHARGES        C2  CRUK    D  DOT         M  PADS      POST  
StockCode                                                                    
10002          0.000000  0.029361   0.0  0.0  0.0  0.059423  0.00  0.070057  
10080          0.000000  0.000000   0.0  0.0  0.0  0.017244  0.00  0.000000  
10120          0.000000  0.000000   0.0  0.0  0.0  0.075165  0.00  0.000000  
10125          0.000000  0.000000   0.0  0.0  0.0  0.075165  0.00  0.071209  
10133          0.000000  0.037139   0.0  0.0  0.0  0.067648  0.05  0.016615  
...                 ...       ...   ...  ...  ...       ...   ...       ...  
D              0.000000  0.000000   0.0  0.0  0.0  0.000000  0.00  0.000000  
DOT            0.000000  0.000000   0.0  0.0  1.0  0.000000  0.00  0.000000  
M              0.071307  0.000000   0.0  0.0  0.0  1.000000  0.00  0.045793  
PADS           0.000000  0.000000   0.0  0.0  0.0  0.000000  1.00  0.000000  
POST           0.017514  0.020569   0.0  0.0  0.0  0.045793  0.00  1.000000  

[3684 rows x 3684 columns]

- Making Recommendations

In [ ]:
top_10_similar_items = list(
    item_item_sim_matrix\
        .loc[23166]\
        .sort_values(ascending=False)\
        .iloc[:10]\
    .index
)

top_10_similar_items

[23166, 23165, 23167, 22993, 23307, 22722, 23243, 22666, 22720, 22961]

In [ ]:
df.loc[
    df['StockCode'].isin(top_10_similar_items),
    ['StockCode', 'Description']
].drop_duplicates().set_index('StockCode').loc[top_10_similar_items]

Description
StockCode                                     
23166           MEDIUM CERAMIC TOP STORAGE JAR
23165            LARGE CERAMIC TOP STORAGE JAR
23167           SMALL CERAMIC TOP STORAGE JAR 
22993             SET OF 4 PANTRY JELLY MOULDS
23307      SET OF 60 PANTRY DESIGN CAKE CASES 
22722        SET OF 6 SPICE TINS PANTRY DESIGN
23243      SET OF TEA COFFEE SUGAR TINS PANTRY
22666          RECIPE BOX PANTRY YELLOW DESIGN
22720        SET OF 3 CAKE TINS PANTRY DESIGN 
22961                   JAM MAKING SET PRINTED

23166 HŨ BẢO QUẢN CERAMIC TOP VỪA
23165 HŨ BẢO QUẢN CERAMIC TOP LỚN
23167 HŨ BẢO QUẢN CERAMIC TOP NHỎ
22993 BỘ 4 KHUÔN THẠCH
23307 BỘ 60 HỘP BÁNH THIẾT KẾ PANTRY
22722 BỘ 6 HỘP GIA VỊ THIẾT KẾ PANTRY
23243 BỘ TRÀ CÀ PHÊ HỘP ĐƯỜNG PANTRY
22666 HỘP CÔNG THỨC THIẾT KẾ MÀU VÀNG
22720 BỘ 3 HỘP BÁNH HIẾT KẾ TIN
22961 BỘ LÀM Mứt ĐƯỢC IN